compiler/rustc_apfloat/src/ieee.rs

```diff
@@ -1511,23 +1511,16 @@ impl<S: Semantics, T: Semantics> FloatConvert<IeeeFloat<T>> for IeeeFloat<S> {
                 sig::set_bit(&mut r.sig, T::PRECISION - 1);
             }
 
-            // If we are truncating NaN, it is possible that we shifted out all of the
-            // set bits in a signalling NaN payload. But NaN must remain NaN, so some
-            // bit in the significand must be set (otherwise it is Inf).
-            // This can only happen with sNaN. Set the 1st bit after the quiet bit,
-            // so that we still have an sNaN.
-            if r.sig[0] == 0 {
-                assert!(shift < 0, "Should not lose NaN payload on extend");
-                assert!(T::PRECISION >= 3, "Unexpectedly narrow significand");
-                assert!(*loses_info, "Missing payload should have set lost info");
-                sig::set_bit(&mut r.sig, T::PRECISION - 3);
-            }
-
-            // gcc forces the Quiet bit on, which means (float)(double)(float_sNan)
-            // does not give you back the same bits. This is dubious, and we
-            // don't currently do it. You're really supposed to get
-            // an invalid operation signal at runtime, but nobody does that.
-            status = Status::OK;
+            // Convert of sNaN creates qNaN and raises an exception (invalid op).
```

Comment:
I don't know what "raises an exception" means here, but wouldn't this mean the cast fails when run in Miri?

Reply:
I think that's the purpose of signaling NaNs. Quoting [Wikipedia](https://en.wikipedia.org/wiki/NaN#Signaling_NaN):

> Signaling NaNs, or sNaNs, are special forms of a NaN that, when consumed by most operations, should raise the invalid operation exception and then, if appropriate, be "quieted" into a qNaN that may then propagate.

This is precisely what we (and LLVM) do here :).

Reply:
It was my understanding that this is generally not implemented, i.e., sNaNs behave like qNaNs with modern compilers on modern hardware? But maybe I misunderstood.

Given that your new testcases passes, it does not seem like this "exception" actually does anything in Miri though, so I guess it's okay.

Reply:
Yeah by default trapping is off, but it can be enabled by e.g. `feenableexcept`. If trapping is off, only some exception flags are set [in a specific "status word" register](http://www.ray.masmcode.com/tutorial/fpuchap1.htm#sword), readable by e.g. [_statusfp](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/status87-statusfp-statusfp2?view=vs-2019) on Windows (couldn't find the unix analog).

Reply:
On x86, it's the [`fstsw` instruction](https://www.felixcloutier.com/x86/fstsw:fnstsw).